### ISSUE TITLE
SQLServer tooooo slow when called from Linux

### ISSUE THREAD

**Issue by CristianSolervicens** [dotnet/runtime#37631]
I have a SqlServer 2012 sp3 under Windows Server 2012. The app is buit in Linux Ubuntu 14.04 with php 5.5.9 and uses php5-sybase library to connect to SqlServer. Both, Linux and Windows are virtual machines running on vmware 6.5 esx.

In the database there are many sp's, but there is one procedure that sometimes, when I call it from linux, it takes 45 sec, and from windows allways no more than 3 sec (app 45000 readings), but from linux it make more than 1 milion readings. The same happens from an ubuntu 14.04 and an ubuntu 20.04, and from php5.6 and php5.6-sybase and I make a small app with python3.4 (Ubuntu 14) and python3.8 with pymssql (Ubuntu 20.04) just to check, and behaves as bad as from php (45 sec. Mo3re than 1M readings).

From php I wrote  "set concat_null_yields_null on" and surprisingly it behaves exactly like from windows, but only for a few hours (2-3 sec. And 45000 readings). When it starts to take 45 secs again I change some other set options and "set concat_null_yields_null off",  and the procedure starts to behave as good as in windows again.

Any ideas about what's happening?, and how to correct it to behave allways as good as it should be?.

From windows I've make the call using Management Studio and from an app I made in c# with Ado.NET. I've got the performance counters (cpu, time and readings) with sql server Profiler.

I modified the proc adding to it the hint "with recompile".
As you guess I installed a new Ubuntu and test the proc. And the application from there, and behaves the same way (the Ubuntu 20.04 I mentioned before). I've modified the proc in 3 or 4 different ways... I have no more ideas to try.

**Comment by Wraith2** [comment]
This doesn't seem like a .net problem, from your description it sounds like the problem is in the php driver for sql server so you'd need to find the maintainers of that and ask them.

**Comment by KevinCathcart** [comment]
Assuming both applications are calling to exactly the same instance of the same database, and are running exactly the same stored procedure, this would indicate that they are using different query plans.

Most likely your php library enables a different set of SET options by default than the .NET client does. This prevents you from using the same cached plan as c# app uses.

The relevant options and their .NET defaults are ANSI_NULL_DFLT_ON, ANSI_NULLS, ANSI_PADDING, ANSI_WARNINGS, CONACT_NULLS_YIELD_NULL, and QUOTED_IDENTIFIER, which all default to ON in the .NET client; and ARITHABORT, which defaults to OFF in the .NET client.

If you change all the SET options to match the C# default, then you would use the same plan from both applications. But, that is not a fix, only an explanation for why they differ. I suspect that if you get the settings to match, you will likely end up experiencing the slower plan in the C# client, at least some of the time.

The actual underlying problem is probably parameter sniffing causing a poor plan to be chosen. Differences in the usage patterns of the php and c# app likely cause different parmeter values to get sniffed, when compiling the procedure. 

The most detailed guide on all of this that I am aware of is: <http://www.sommarskog.se/query-plan-mysteries.html>, which not only discusses the impact of these settings on the plan cache, but also why and how parameter sniffing can become a problem, and the various things you can try to fix it, as well as a few other possible causes of basically this same issue that are not parameter sniffing related.

If that article does not help you fix your issue, you probably need to work with either Microsoft support (a paid SQL Server support ticket) or work with an independent SQL Server performance consultant.

**Comment by Clockwork-Muse** [comment]
I agree, parameter sniffing seems like a good first thing to check.  [Option `RECOMPILE`](https://stackoverflow.com/questions/20864934/option-recompile-is-always-faster-why) _may_ be another thing to try, along with updating statistics and similar work.

**Comment by CristianSolervicens** [comment]
Hello,

My first thought was: There is a problem with the "open" sql server
connector buit for PHP, it's quite old, but when the same happened with the
python connector I thought the problem is with Linux itself.

Now I think you're right, tonight I dreamed about the issue, and I got the
same explanation "Execution Plan", I even thought about the
possibillitie of the impact of the string encoding, or string termination
in the different environments and that might result in a different hash for
the "same" query and therefore different execution plans.
Today the application changes their behavior, and when I called the
procedure from the C# app with the most used parameters it also took 45
seconds to respond. So I think you'r right, the problem is the execution
plan for the query. So, now I have to figure out what's in my procedure
that sometimes it produces this awful execution plan.

Thank you all !!!


Cristian Solervicéns C.
Cel: 99433 23 69

**Comment by CristianSolervicens** [comment]
> I agree, parameter sniffing seems like a good first thing to check. [Option `RECOMPILE`](https://stackoverflow.com/questions/20864934/option-recompile-is-always-faster-why) _may_ be another thing to try, along with updating statistics and similar work.

Thansk, the hint "WITH RECOMPILE" was the first change I made to the stored procedure (didn't work).

**Comment by cheenamalhotra** [comment]
@CristianSolervicens 

> So, now I have to figure out what's in my procedure that sometimes it produces this awful execution plan.

If you are using server-side pagination by using ORDER BY clause, try to change that to something else to paginate or use client-side pagination instead. Use minimalistic queries to investigate what's causing the lag. 

You can also try to enable [AUTO UPDATE STATISTICS ASYNCHRONOUSLY](https://docs.microsoft.com/en-us/sql/relational-databases/statistics/statistics?view=sql-server-ver15#auto_update_statistics_async) if you see time lag during clustered index creation, so that it does not update your table statistics during query execution.

Or try running [AdaptiveIndexDefrag](https://github.com/Microsoft/tigertoolbox/tree/master/AdaptiveIndexDefrag) on your database to automate index and statistics management.

**Comment by CristianSolervicens** [comment]
I've solved the problem. It was parameter sniffing problem, that sometimes
result into a bad execution plan.

The person who wrote the procedure (it's long but no too long) used two
functions and one view.
Both functions have optional parameters and the parameters were used in the
where clause doing something like:

and (  @param IS NULL Or Column = @param)

Because of this the optimizer can't do it's job correctly, and the
execution plan sometimes include an indexscan.
So, I took the code of both of the functions and I put it into the
procedure deleting the trash parameters that I don't need, I did
something similar with the view, adding the actual tables that are needed
to get the result.

Before the fix a "good" response time was 2.5 sec. but when the optimizer
choosed the indexscan the time rised up to 45 sec. and a bit more than 1
million reads!.
After the fix, the result is a stable and consistent execution plan, the
execution time is 0.83 sec. and the reads are between 30k-45k (depending
the parameters).

*I really appreciate all the responses, for the time spent and the
goodwill.*


Cristian Solervicéns C.
Cel: 99433 23 69

**Comment by Clockwork-Muse** [comment]
... yes, [this was an expected possibility](https://stackoverflow.com/a/59047547/812837).

> Because of this the optimizer can't do it's job correctly, and the
execution plan sometimes include an indexscan.
So, I took the code of both of the functions and I put it into the
procedure deleting the trash parameters that I don't need, I did
something similar with the view, adding the actual tables that are needed
to get the result.

.... I'd be careful of this, however, as now you have a place that should likely be updated if the original functions or view changes.  Also, you've solved the issue for the singular query you were working on, but it's possible that anything else using those functions/view is also experiencing similar slowdowns.  It may be worth it to find the references to them to figure out whether or not they are (if not, why not?), and possibly update them to a more performant version (and then changing your procedure to use them again).
